src/dodal/beamlines/beamline_utils.py

```diff
@@ -30,21 +36,44 @@ def list_active_devices() -> List[str]:
     return list(ACTIVE_DEVICES.keys())
 
 
-def active_device_is_same_type(active_device, device):
+def active_device_is_same_type(
+    active_device: AnyDevice, device: Callable[..., AnyDevice]
+) -> bool:
     return type(active_device) == device
 
 
+def _wait_for_connection(
+    device: AnyDevice, timeout: float = DEFAULT_CONNECTION_TIMEOUT
+) -> None:
+    if isinstance(device, OphydV1Device):
+        device.wait_for_connection(timeout=timeout)
+    elif isinstance(device, OphydV2Device):
+        call_in_bluesky_event_loop(
+            asyncio.wait_for(
+                v2_device_wait_for_connection(coros=device.connect()),
+                timeout=timeout,
+            )
+        )
```

Comment:
@RAYemelyanova or @evalott100 might know more

Reply:
After chatting to @Tom-Willemsen decided to pull this into an ophyd issue that asks what the idomatic way of doing it is and leave this here.

Reply:
Key thing discussed was the existing `DeviceCollector` implementation isn't really suitable for us as it's API is based on walking stack frames of the caller (!) and looking for new locals (!). Seems to be mostly targeted at REPL use currently?

We *could* do something like

```python
with DeviceCollector():
    _ = device
```

But that feels really wrong as it looks like it doesn't do anything, and will probably break during refactors. There's also  peculiarities like

```python
with DeviceCollector():
    for device in [...]:
        _ = device
```

Which would only connect the last device. I'm just not really happy with laying these sorts of traps in our dodal code, it would be nice to have a simple/standard API in ophyd with a normal function we can just call passing in one or more devices to ensure they're connected, without any stack-walking or dependence on caller locals etc.

Reply:
I agree with this; in my review I also commented the same thing. I'll make sure to bring it up in the next ophyd v2 meeting with NSLS2 (which will be next Tuesday).